### PR TITLE
layers: Make PIPELINE_STATE data private

### PR DESCRIPTION
**Pull request by ncesario-lunarg**
Make a number of fields in PIPELINE_STATE private. This is done to
enforce outside code to use accessors, which is a necessary precursor to
breaking up pipeline state into sub-states.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -1622,7 +1630,7 @@ bool CoreChecks::ValidatePipelineLocked(std::vector<std::shared_ptr<PIPELINE_STA
         }
 
         // Validate polygon mode
-        auto raster_state_ci = create_info.pRasterizationState;
+        auto raster_state_ci = pipeline->RasterizationState();
```
Comment (jeremyg-lunarg) [inline layers/core_validation.cpp]: auto *

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -1964,63 +1980,60 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
...
                              "enabled. It is invalid.",
-                             pipelineIndex, string_VkPrimitiveTopology(create_info.pInputAssemblyState->topology));
+                             pipelineIndex, string_VkPrimitiveTopology(ia_state->topology));
         }
     }
 
     // If a rasterization state is provided...
-    if (create_info.pRasterizationState) {
-        if ((create_info.pRasterizationState->depthClampEnable == VK_TRUE) && (!enabled_features.core.depthClamp)) {
+    const auto raster_state = pPipeline->RasterizationState();
```
Comment (jeremyg-lunarg) [inline layers/core_validation.cpp:2034]: const auto *?

File: layers/synchronization_validation.cpp [hunk 3]
```diff
@@ -1850,8 +1850,8 @@ bool CommandBufferAccessContext::ValidateDispatchDrawDescriptorSet(VkPipelineBin
     using TexelDescriptor = cvdescriptorset::TexelDescriptor;
 
     for (const auto &stage_state : pipe->stage_state) {
-        if (stage_state.stage_flag == VK_SHADER_STAGE_FRAGMENT_BIT && pipe->create_info.graphics.pRasterizationState &&
-            pipe->create_info.graphics.pRasterizationState->rasterizerDiscardEnable) {
+        const auto raster_state = pipe->RasterizationState();
```
Comment (jeremyg-lunarg) [inline layers/synchronization_validation.cpp:1853]: const auto *?

File: layers/synchronization_validation.cpp [hunk 4]
```diff
@@ -1987,8 +1987,8 @@ void CommandBufferAccessContext::RecordDispatchDrawDescriptorSet(VkPipelineBindP
...
```
Comment (jeremyg-lunarg) [inline layers/synchronization_validation.cpp:1990]: const auto *?

File: layers/synchronization_validation.cpp [hunk 5]
```diff
@@ -2332,8 +2332,8 @@ bool RenderPassAccessContext::ValidateDrawSubpassAttachment(const CommandExecuti
         return skip;
     }
 
-    const auto &create_info = pipe->create_info.graphics;
-    if (create_info.pRasterizationState && create_info.pRasterizationState->rasterizerDiscardEnable) {
+    const auto raster_state = pipe->RasterizationState();
```
Comment (jeremyg-lunarg) [inline layers/synchronization_validation.cpp:2335]: const auto*?

File: layers/core_validation.cpp [hunk 6]
```diff
@@ -632,9 +632,9 @@ bool CoreChecks::ValidateRenderPassCompatibility(const char *type1_string, const
 
 // For given pipeline, return number of MSAA samples, or one if MSAA disabled
 static VkSampleCountFlagBits GetNumSamples(PIPELINE_STATE const *pipe) {
-    if (pipe->create_info.graphics.pMultisampleState != NULL &&
-        VK_STRUCTURE_TYPE_PIPELINE_MULTISAMPLE_STATE_CREATE_INFO == pipe->create_info.graphics.pMultisampleState->sType) {
-        return pipe->create_info.graphics.pMultisampleState->rasterizationSamples;
+    const auto ms_state = pipe->MultisampleState();
```
Comment (jeremyg-lunarg) [inline layers/core_validation.cpp:635]: const auto*?

File: layers/core_validation.cpp [hunk 7]
```diff
@@ -1358,7 +1361,7 @@ bool CoreChecks::ValidateCmdBufDrawState(const CMD_BUFFER_STATE *cb_node, CMD_TY
     }
     // Now complete other state checks
     string error_string;
-    const auto *pipeline_layout = pipe->pipeline_layout.get();
+    auto const &pipeline_layout = pipe->PipelineLayoutState();
```
Comment (jeremyg-lunarg) [inline layers/core_validation.cpp]: this one is  backwards.

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM. We really should figure out an automated way to find and enforce the `auto *`for plain pointers and `const` ordering stuff.

File: layers/core_validation.cpp (hunk 1 above)
Comment (ncesario-lunarg) [reply layers/core_validation.cpp]: I don't think this is necessary for pointers. If a pointer type is returned, `auto` is deduced as a pointer. I think the issue is references, which `auto` will never do implicitly. e.g., https://godbolt.org/z/7TWT9Tzah

Notice the types for a1 and a2 are equivalent (namely, `Pi` meaning "pointer to int").

**Pull request closed without merging**